asf_installer_survey/notebooks/section_6_analysis.py

```diff
@@ -0,0 +1,1670 @@
+
+# %% [markdown]
+# ### RQ 6.04 | How does the offer of repairs, servicing and maintenance for all heat pump
+# systems or just ones survey participants have installed differ depending on the size
+# of the business they work for or own?
+# - SQ: 43a, 43b, 43c
+# - Sample:
+# - Company_size_owner
+# - Company size_employee
+
+# %%
+# CompanySizeOwner V2
+# Generate crosstab dataframe
+df_604_crosstab_owner_V2 = crosstable("CompanySizeOwnerV2",
+                                        data,
+                                        "SQ43",
+                                        ans_43
+                                        )
+
+# Save to .csv and display
+df_604_crosstab_owner_V2.to_csv("../../outputs/section6/csv/604_owner_V2.csv")
+df_604_crosstab_owner_V2
+
+# %%
+# Generate figure
+groupedbar(df_604_crosstab_owner_V2,
+           "604_owner_V2",
+           "SQ 43a. & 43b. I/my business offers heat pump repairs, servicing and maintenance:"
+           )
+
+# %%
+# Generate figure
+donut(df_604_crosstab_owner_V2,
+           "604_owner_V2",
+           "SQ 43a. & 43b. I/my business offers heat pump repairs, servicing and maintenance:"
+           )
+
+# %%
+# CompanySizeEmployee V2
+# Generate crosstab dataframe
+df_604_crosstab_employee_V2 = crosstable("CompanySizeEmployeeV2",
+                                         data,
+                                        "SQ43",
+                                        ans_43
+                                        )
+
+# Save to .csv and display
+df_604_crosstab_employee_V2.to_csv("../../outputs/section6/csv/604_employee_V2.csv")
+df_604_crosstab_employee_V2
+
+# %%
+# Generate figure
+groupedbar(df_604_crosstab_employee_V2,
+           "604_employee_V2",
+           "SQ 43c. The business I work for offers heat pump repairs, servicing and maintenance:"
+           )
+
+# %%
+# Generate figure
+donut(df_604_crosstab_employee_V2,
+           "604_employee_V2",
+           "SQ 43c. The business I work for offers heat pump repairs, servicing and maintenance:"
+           )
+
+# %% [markdown]
+# ### RQ 6.05 | Why do survey participants who only offer repairs, servicing and maintenance for heat pump systems they’ve installed choose not to do so for all other systems?
+# - SQ: 44a, 44b
+# - Sample
+# - Standard_output
+# <p>
+# N.B. Select all that apply type question
+# <p>
+# Update: Free text responses recoded on 2024-04-17
+
+# %%
+# Print possible answers and check counts
+data[col.q44a[0]].value_counts()
+data[col.q44b[0]].value_counts()
+
+# %%
+def transform_44(x):
+    """
+    Function to be applied to dataframe which separates answers selected
+    by creating an individual tally column for each possible answer. 
+
+    Args:
+        x: pandas dataframe containing responses from analytical sample.
+    """
+    
+    # Tally occurrence of each answer
+    if response in x[col.q44a[0]] or response in x[col.q44b[0]]:
+        return response
+    else:
+        return None
```

Comment:
It might be helpful to raise an exception or log a warning in this case, to make it easier to debug if unexpected values are encountered!

Reply:
Great tip! Added accordingly:

```
    # Tally occurrence of each answer
    if response in x[col.q44a[0]] or response in x[col.q44b[0]]:
        return response
    elif "Not asked" in x[col.q44a[0]] or "Not asked" in x[col.q44b[0]]:
        return None
    else:
        raise ValueError(f"Unexpected response : {x[col.q44a[0]]},{x[col.q44b[0]]}")
```